src/webgpu/shader/validation/expression/call/builtin/subgroupAdd.spec.ts

```diff
@@ -0,0 +1,194 @@
+export const description = `
+Validation tests for subgroupAdd and subgroupExclusiveAdd
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import { Type, elementTypeOf, kAllScalarsAndVectors } from '../../../../../util/conversion.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kBuiltins = ['subgroupAdd', 'subgroupExclusiveAdd'] as const;
+
+const kStages: Record<string, (builtin: string) => string> = {
+  constant: (builtin: string) => {
+    return `
+enable subgroups;
+@compute @workgroup_size(16)
+fn main() {
+  const x = ${builtin}(0);
+}`;
+  },
+  override: (builtin: string) => {
+    return `
+enable subgroups;
+override o = ${builtin}(0);`;
+  },
+  runtime: (builtin: string) => {
+    return `
+enable subgroups;
+@compute @workgroup_size(16)
+fn main() {
+  let x = ${builtin}(0);
+}`;
+  },
+};
+
+g.test('early_eval')
+  .desc('Ensures the builtin is not able to be compile time evaluated')
+  .params(u => u.combine('stage', keysOf(kStages)).beginSubcases().combine('builtin', kBuiltins))
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('subgroups' as GPUFeatureName);
+  })
+  .fn(t => {
+    const code = kStages[t.params.stage](t.params.builtin);
+    t.expectCompileResult(t.params.stage === 'runtime', code);
+  });
+
+g.test('must_use')
+  .desc('Tests that the builtin has the @must_use attribute')
+  .params(u =>
+    u
+      .combine('must_use', [true, false] as const)
+      .beginSubcases()
+      .combine('builtin', kBuiltins)
+  )
+  .beforeAllSubcases(t => {
+    t.selectDeviceOrSkipTestCase('subgroups' as GPUFeatureName);
+  })
+  .fn(t => {
+    const wgsl = `
+enable subgroups;
+@compute @workgroup_size(16)
+fn main() {
+  ${t.params.must_use ? '_ = ' : ''}${t.params.builtin}(0);
+}`;
+
+    t.expectCompileResult(t.params.must_use, wgsl);
+  });
+
+const kArgumentTypes = objectsToRecord(kAllScalarsAndVectors);
+
+g.test('data_type')
```

Comment:
Should also test that a struct and array are not allowed, and e.g. the result of frexp (a builtin struct)

Reply:
heh, pointer too.